### PR TITLE
fix(mcp,api): audit logging, OAuth scope parity, Prisma 7 constraint fields and dry-run scopes

### PR DESCRIPTION
**Pull request by xernobyl**
## What & why

Five defects found during the 5.3 release QA (§9 API/MCP and §5 Surveys). They share no code, but each is a case where the product's documented behaviour and its actual behaviour disagree, and each fix is small — so they go together rather than as five branches.

**Duplicate creates returned 500 instead of 409.** `@prisma/adapter-pg` builds the P2002 column list by regex-scraping the Postgres error DETAIL and never unquotes it, so `singleUseId` arrives as `"singleUseId"`. Every caller compares with exact equality, so none matched and the error fell through to a generic `DatabaseError`. Fixed once in `getUniqueConstraintFields`, which repairs four call sites — including one nobody had noticed: **re-adding an existing tag to a response was returning a 500** instead of being an idempotent no-op.

**Every new MCP client failed its first OAuth connect.** The 401 `WWW-Authenticate` challenge advertised six scopes; the protected-resource metadata advertises those plus `offline_access`. A client that hits the 401 before fetching the metadata uses the challenge as its registration scope, registers narrow, authorizes wide, and the provider rejects it with `invalid_scope`. It succeeds on retry, once the metadata is cached. Both lists now derive from one constant.

**No MCP mutation has ever been audited.** The tools passed the bare path `/api/mcp` as the audit event's `apiUrl`; the schema validates it with `z.url()` and `logAuditEvent` catches the failure and downgrades it to a `logger.error`, dropping the event. That includes `delete_feedback_record`, which the tooling documents as producing an audit entry containing the removed content.

**Two dry-run tools were unusable by read-scoped callers.** `validate_survey` and `test_workflow` are registered `:read` and annotated `readOnlyHint`, but their handlers authorized at `readWrite`, so a read-scoped token got a 403 from a tool that writes nothing. Both handlers now gate at `read`. The check still runs — only the level drops.

`MCP_API_ROUTE` and the new `MCP_AUDIT_API_URL` are kept as separate constants: the first is the RFC 9457 problem-document `instance` member, where a relative reference is correct, and conflating the two is what caused the audit bug.

## Linear ticket

- Fixes ENG-2174 — Prisma 7 quoted constraint fields
- Fixes ENG-2175 — MCP 401 challenge / metadata scope mismatch
- Fixes ENG-2173 — MCP mutations emit no audit log
- Fixes ENG-2179 — `validate_survey` scope mismatch
- Fixes ENG-2223 — `test_workflow`, same mismatch (found while fixing ENG-2179)

## How this was tested

**Automated** — all green on the final state:

- `apps/web` full suite: **590 files / 7604 tests passed** ✅
- `apps/web` integration suite (real Postgres): **21 files / 77 tests passed** ✅
- `@formbricks/workflows`: **13 files / 242 tests passed** ✅
- `pnpm typecheck` (apps/web): **0 errors** ✅
- `eslint` on every changed file: **clean** ✅

**Rebased onto #8718**, which comma-separated the `WWW-Authenticate` auth-params per RFC 9110 §11.6.1. The two PRs touch the same challenge string, so this branch reconciles them: the header is now `Bearer resource_metadata="…", scope="…"` with all seven scopes, and I re-verified both fixes hold together by parsing the live header with a strict `#auth-param` parser — `resource_metadata` is extracted correctly and the scope list has seven entries.

That rebase also had to repair two assertions. #8718 replaced two robust checks in `mcp/auth.test.ts` with exact-match literals of the whole header, which this branch's extra scope would have silently invalidated. They now match the RFC grammar (`/^Bearer resource_metadata="[^"]+", scope="[^"]+"$/`) plus the metadata URL each case is actually about, so the comma stays guarded without re-pinning the scope list; the full value is still pinned exactly once, from the real constant, in `app/api/mcp/route.test.ts`.

**Verified over HTTP** against a dev server running this branch, with the same checks run first against `release/5.3` to capture the "before":

| Check | release/5.3 | this branch |
|---|---|---|
| 401 challenge vs metadata `scopes_supported` | differ by `offline_access` | identical |
| DCR round-trip — register with the challenge scope | client lacks `offline_access` | covers every advertised scope |
| Duplicate `singleUseId` on `/api/v1/client/{ws}/responses` | 200 then **500** | 200 then **409** `{"code":"conflict",…}` |
| Read-only key → `POST /api/v3/surveys/validate` | 403 | **200**, `valid:true` |
| Read-only key → `POST /api/v3/surveys` (control) | 403 | **403** — writes still refused |

Reproducible with [`qa-mcp-scope-check.sh`](https://github.com/formbricks/formbricks/pull/8743) style checks; the scope one is a single `diff` of the two scope lists.

Still **not** observed end-to-end: the MCP browser OAuth connect (needs an interactive login, and a genuinely de-registered client — a cached DCR registration hides the bug entirely), and the audit-log line (needs `AUDIT_LOG_ENABLED=1` plus a log tail). `test_workflow` could not be exercised over HTTP either: workflows are licence-gated, and the local instance answers `403 "Workflows are not enabled for this organization"` before authorization is reached. Its level change is covered directly by a unit assertion (`expect(authorizeAllow).toHaveBeenCalledWith(workspaceId, "read")` immediately after `testWorkflow`), with a table-driven control over all eight write handlers proving they stay at `readWrite`. All three remain a plan for a tester under "QA / Test Plan".

**Coverage on the changed surface** — 95.08% statements / 83.38% branches / 97.7% functions / 95.31% lines. The new helper in `prisma-constraint.ts` is **100%** (17/17 statements, 11/11 branches, 5/5 functions).

**Tests added**

- `prisma-constraint.test.ts` — quoted composite and single keys, a mixed list (proving `fields[0]` readers are unaffected), `@map()`ed snake_case left alone, the legacy `meta.target` shape, and the outer-pair-only edges (`"`, `""`, `"a"b"`).
- `oauth-protected-resource/route.test.ts` — asserts the 401 challenge scope equals the metadata `scopes_supported` **exactly**, reading the real constants. This is the test that would have caught ENG-2175.
- `mcp/auth.test.ts` — a token scoped **only** to `offline_access` is refused. Nothing asserted this before, and it is the regression the scope fix must not introduce.
- `audit-logs/service.test.ts` — proves a bare-path `apiUrl` drops the event while an absolute one does not, and ties the MCP constant to the audit validator so they cannot drift apart again.
- `workflows.handlers.test.ts` — asserts `testWorkflow` authorizes at `read`.
- **Two integration tests against real Postgres**, because the unit fixtures build the P2002 meta by hand and every one of them passed unquoted names — which is precisely why this shipped. `response-error.integration.test.ts` drives a genuine duplicate `(surveyId, singleUseId)` and a duplicate `displayId`; `prisma-constraint.integration.test.ts` gains a duplicate `Workspace (organizationId, name)`. Both assert the **raw** adapter meta carries the quotes before checking the normalised output, so they cannot pass vacuously.

I confirmed these fail without the fix: disabling the unquoting turns both `response-error` cases into `expected error to be instance of UniqueConstraintError / InvalidInputError` — i.e. the 500 the ticket describes.

**Root cause, from a real Postgres instance:**

```
quote_ident('surveyId')  -> "surveyId"      quote_ident('token_hash') -> token_hash
DETAIL: Key ("surveyId", "singleUseId")=(s1, u1) already exists.
adapter's regex parse:  ["\"surveyId\"", "\"singleUseId\""]   <- quoted, never matched
snake_case equivalent:  ["token_hash"]                        <- bare, always matched
```

That asymmetry is why the action-class callers (whose lowercase column sorts first) kept working while the response endpoints did not.

**Fixtures**: reused the existing `uniqueViolation` / `adapterP2002` / `legacyP2002` builders rather than adding another — the repo already has nine hand-rolled copies. Consolidating those, and the four parallel `isUniqueConstraintError` definitions, is tracked on ENG-1809 (which already exists for this surface and now carries the findings from here) rather than being buried in this diff.

## Breaking changes

None

## QA / Test Plan

**How to test**

- [ ] **Duplicate single-use response** — open a link survey with single-use links enabled and submit the same single-use link twice. Second submit → **409** with body `{"code":"conflict","message":"Response already submitted for this single-use link"}`. Previously **500**. Check both `/api/v1/client/{workspaceId}/responses` and `/api/v2/...`.
- [ ] **Duplicate tag** — open a response, add a tag, then add the **same** tag again. Expect a silent no-op. Previously an error toast / 500.
- [ ] **MCP first connect** — revoke any existing client at `/account/settings/authorized-apps`, then connect a fresh MCP client (e.g. `claude mcp add --transport http <name> <host>/api/mcp`, then `/mcp` → authenticate). Expect the **first** attempt to succeed. Previously it failed with `invalid_scope` and only worked on retry.
- [ ] **Challenge matches metadata** — `GET /.well-known/oauth-protected-resource/api/mcp` and an unauthenticated `POST /api/mcp`; the `scopes_supported` array and the `scope="…"` in the 401 `WWW-Authenticate` header must list the same seven scopes, `offline_access` included. The header must also keep the **comma** between the two auth-params (`resource_metadata="…", scope="…"`) that #8718 introduced — a space there is malformed and hides `resource_metadata` from strict client parsers.
- [ ] **MCP audit** — with `AUDIT_LOG_ENABLED=1`, run any MCP mutation (e.g. `create_survey`) and check the application log for a `"level":"audit"` record with `"apiUrl":"<host>/api/mcp"`. Previously there was no audit record, only a `Failed to log audit event` error line.
- [ ] **Read-scoped dry runs** — with an API key that has **read** permission on the workspace: `POST /api/v3/surveys/validate` → **200**, and `POST /api/v3/workflows/{id}/test` → **200**. Both were **403**.
- [ ] **Negative controls (must still be 403)** — same read-only key against `POST /api/v3/surveys` and `PATCH /api/v3/workflows/{id}`; and against a survey/workflow in a workspace the key has no access to. If any of these returns 200, the authorization change went too far.

**Preconditions / test data**

- A workspace with single-use links enabled on a link survey, plus at least one response and one tag.
- Two API keys on the same workspace: one **read**, one **manage**.
- `AUDIT_LOG_ENABLED=1` and access to the application log for the audit check.
- An MCP client that can be fully de-registered (revoking at `/account/settings/authorized-apps` is required — a cached registration masks the connect fix).

**Risks & regressions**

- The scope change is the highest-risk part: `offline_access` must reach the 401 challenge and **nothing else**. The baseline authentication gate stays on resource scopes, so a token holding only `offline_access` is still refused — worth re-checking if the auth code is touched again.
- The constraint fix affects every P2002 consumer, not just the response endpoints. Action-class duplicate names/keys should still report the right field — they read `fields[0]` and their lowercase column sorts first, so they were unaffected either way.
- The dry-run authorization change also applies to the plain REST endpoints, so read-only API keys and read-only team members can now call them. Neither branch writes, and a `:read` caller could already obtain the same information from `get_survey` / `get_workflow`.
- Self-hosters with `AUDIT_LOG_ENABLED=1` will start receiving audit events for MCP mutations that were previously dropped. That is the fix, but it is a volume change worth expecting.
- A feedback-source mapping collision now reports `FEEDBACK_SOURCE_FORMBRICKS_MAPPING_DUPLICATE` instead of `FEEDBACK_SOURCE_NAME_DUPLICATE` — same 400, same error class, but a visible message change. The previous classification was simply wrong.

**Migrations / env / cutover**

- none

### CODE REVIEW

File: apps/web/modules/mcp/auth.ts [hunk 1]
```diff
@@ -370,14 +361,24 @@ async function authenticateMcpOAuthBearer(
...
+  // `scope` names the scopes *required* for the resource, and `offline_access` is not one of them.
+  if (!hasAnyMcpScope(authInfo, MCP_RESOURCE_SCOPES)) {
     log.warn({ statusCode: 403, clientId: authInfo.clientId }, "MCP OAuth token missing every MCP scope");
     return {
       ok: false,
       requestId,
       response: withInsufficientScopeChallenge(
         problemForbidden(requestId, "OAuth token does not include the required MCP scope", instance),
-        [...MCP_MINIMUM_SCOPES]
+        [...MCP_RESOURCE_SCOPES]
```
[the remark this answers is not in the document]
Comment (xernobyl) [reply apps/web/modules/mcp/auth.ts:384]: Valid catch — fixed in a8bd95c6c.

The handbook was the last reference to the alias. I updated the code comment in `workspaces.ts` that named it but missed this one; a repo-wide grep now returns zero hits across `apps/`, `packages/` and `docs/`.

Went with `MCP_RESOURCE_SCOPES` rather than dropping the identifier, because the surrounding paragraph is explaining *why* `list_workspaces` gates on any resource read scope, and naming the constant the gate actually reads keeps that traceable. The baseline itself is unchanged — at least one resource scope.

Worth noting for anyone reading the diff: the alias was removed deliberately, not just tidied away. It sat two lines above the challenge constant and its name ("minimum") invited pointing it at the protected-resource list, which would let a token holding only `offline_access` — no resource access at all — authenticate to the MCP server. `auth.ts` now carries a comment saying so, and there is a new test asserting an `offline_access`-only token is still refused.

**Review by mattinannt — COMMENTED** [review]
## Review summary

This review was conducted by the Claude Code PR review plugin (five specialist passes: general code review, test coverage, silent-failure hunting, comment accuracy, plus manual verification of the top findings). Every claim below was traced against the code on this branch rather than taken from the PR description.

**Verdict: the five fixes are correct, but there is one blocker.** A latent bug in the batch feedback-record path becomes live *because* of the `apiUrl` fix, and several newly-written comments and doc lines state things the code does not do.

Both authorization downgrades were independently traced and confirmed genuinely write-free:

- `validateV3Survey` — the create branch is a zod parse plus identity (`prepare.ts:181-189`); the patch branch builds a document from the loaded row and merges in memory (`prepare.ts:191-219`). Both are synchronous, so no I/O is possible. The only prisma-importing dependency, `languages.ts`, has its write inside `ensureV3WorkspaceLanguages`, which `validateV3Survey` never calls — it calls only the pure `deriveV3SurveyLanguageRequests`.
- `testWorkflow` — one read via `loadAndAuthorize`, then `safeParse` plus two verification reads. No run row, no enqueue, no `recordAuditSafely` (contrast the `enable` handler directly above it).

Neither returns anything a `read` caller cannot already obtain from `get_survey` / `get_workflow`, so the scope decision itself is sound. CI is green and the new integration tests do run in CI (`.github/workflows/integration-tests.yml`).

---

## Critical

### 1. `create_feedback_records` now writes fabricated "success" audit records

`apps/web/modules/mcp/tools/feedback-records.ts:252`

The batch path filters for the entries the operation stamped:

```ts
const stamped = auditLogs.filter((auditLog) => auditLog?.targetId);
```

`buildAuditLogBaseObject` seeds `targetId: UNKNOWN_DATA` (`apps/web/app/lib/api/with-api-logging.ts:373`), and `UNKNOWN_DATA === "unknown"` — truthy. **The filter never filters anything.** For a batch of N records:

- **Partial failure** — all N entries pass, all N get `status = "success"`, all N are queued. Records the Hub rejected produce audit events asserting a creation that never happened, with `targetId: "unknown"` and no `newObject`.
- **Total failure** — `stamped.length > 0` still holds, so the function returns at line 259 and the `// Nothing created` branch (lines 261-266) is unreachable dead code. A batch that created nothing emits N success events instead of one failure event.

The filter is pre-existing on `main` (same line). What changes here is that these events previously died inside `logAuditEvent`'s `z.url()` check — **fixing `apiUrl` is what lets them reach the audit sink for the first time.** For the compliance use case that motivates ENG-2173, a false audit record is worse than a missing one, so this PR should not ship without the guard.

`operations.ts:569` documents the intended contract that `UNKNOWN_DATA` breaks: *"records actually created come back with `targetId` set; the caller queues those as successes."*

The tests cannot catch it because the mock returns `{ status: "failure" }` with **no `targetId`** (`feedback-records.test.ts:468`), diverging from production in exactly the field the code branches on. All three batch tests pass green while asserting behaviour production does not have.

Suggested fix — stop inferring "was created" from a field pre-seeded to a truthy sentinel:

```ts
const stamped = auditLogs.filter(
  (auditLog) => auditLog && auditLog.targetId !== UNKNOWN_DATA
);
```

plus a regression test that uses the real `buildV3AuditLog`, or at minimum a mock returning `{ status: "failure", targetId: UNKNOWN_DATA }`.

---

## Important

### 2. Three doc statements now contradict this PR's own changes

This PR's thesis is that documented and actual behaviour must agree, so these should not be left behind:

- `docs/development/technical-handbook/mcp-server.mdx:647` — "create validation still checks workspace **write** access when `workspaceId` is present." Directly contradicted by the table row this PR adds at line 162. Same file the PR edits.
- `docs/api-v3-reference/src/openapi.yml:130-133` — "Dry-run tests create real run records with `isDryRun: true` that mock action execution… `POST .../test` returns a `queued` run to poll." The handler creates no run and returns `{ workflowId, ok, problems }` synchronously. The per-path spec (`src/paths/api_v3_workflows_{workflowId}_test.yml:8`) gets it right, so the two halves of the public API reference disagree — and the overview asserts the opposite of the "persists nothing" premise this PR's downgrade rests on.
- `docs/development/technical-handbook/mcp-server.mdx:809` — "Only **enabled or disabled** workflows can be tested — a draft or archived workflow is rejected." The handler rejects only archived, and its own comment says drafts are testable on purpose.

### 3. "A superset is not enough; the two lists must match exactly" is not true

`apps/web/modules/auth/lib/oauth-urls.ts:83`

The provider's authorize check is a subset test, not equality (`@better-auth/oauth-provider/dist/index.mjs:3872-3877`):

```js
const validScopes = new Set(client.scopes ?? opts.scopes);
const invalidScopes = requestedScopes.filter((scope) => !validScopes?.has(scope));
```

DCR stores the requested scope verbatim and admits any subset of `clientRegistrationAllowedScopes`, so the real invariant is **challenge ⊇ metadata**. A strict superset registers wider, authorizes narrower, and passes both gates; only a *narrower* challenge breaks — which is exactly the ENG-2175 mechanism the comment correctly describes one sentence earlier. The handbook Note at `mcp-server.mdx:256` already phrases this correctly ("not a subset"); align the code comment to the doc rather than the reverse.

### 4. "mock-executes with side effects suppressed" describes a mechanism that does not exist

`packages/workflows/src/handlers/workflows.handlers.ts:508` (new), `docs/development/technical-handbook/mcp-server.mdx:169` (new), `apps/web/modules/mcp/tools/workflows.ts:245`

`testWorkflow` parses the definition and performs three lookups. There is no execution, mocked or otherwise, and nothing to "suppress." This matters beyond pedantry: `isDryRun` is a real persisted concept elsewhere in this codebase (`packages/workflows/src/services/ports.ts:55`), so a maintainer will go looking for a suppression path and land on a stale design — and describing a **`read`-gated** endpoint as "executing" actions invites the wrong conclusion in the next security review of this scope decision.

The accurate wording is already in the JSDoc immediately above (`workflows.handlers.ts:497-505`): *"a pure pre-flight check."* Suggest "validates the definition and resolves its trigger references; no run is created and no side effect is performed."

### 5. Nine of the ten new audit assertions are self-referential — the antipattern this PR removes elsewhere

`expect(buildV3AuditLog).toHaveBeenCalledWith(..., MCP_AUDIT_API_URL)` appears in `feedback-records.test.ts`, `surveys.test.ts` and `workflows.test.ts`. Test and production import the same constant, so if it regressed to `"/api/mcp"` tomorrow every one of those assertions would still pass — reintroducing ENG-2173 with a green suite.

`service.test.ts:94` ("accepts the apiUrl the MCP tools actually pass") is the one genuine guard, because it runs the real value through the real schema. Worth noting that this PR *correctly removes* this same pattern from `auth.test.ts` with the note "re-declaring them here would assert the mock against itself and mask scope drift" — the lesson just was not applied to the audit assertions.

Related: `MCP_API_ROUTE` and `MCP_AUDIT_API_URL` are now both imported into `feedback-records.ts:19` — similarly named, one of which silently destroys audit events. A branded type (`toAuditApiUrl()` returning `string & { __auditApiUrl }`) would make passing the path a compile error, if that seems worth the ceremony.

### 6. Missing negative controls on the two authorization levels this PR lowered

The PR's own QA plan lists these as the must-still-403 controls, so they belong in the suite:

- `packages/workflows/src/handlers/workflows.handlers.ts` gates eight write handlers at `"readWrite"`, but the test file asserts an authorize level in only three places and **all three assert `"read"`**. Lowering `patch` or `delete` to `"read"` — the precise move made twice in this PR — passes the entire suite. A single `test.each` over the eight handlers closes it.
- `createV3SurveyResponse` (`apps/web/app/api/v3/surveys/lib/operations.ts:288`) never asserts its access level; the only `"readWrite"` assertions in `operations.test.ts` are for patch and archive. Since this PR moved `validateV3Survey` across the same `requireV3WorkspaceAccess` seam, create is the obvious next accident.

---

## Suggestions

- **The two silently-fixed P2002 consumers have no regression coverage.** `lib/tagOnResponse/service.ts:35` (the re-tag 500 the description highlights) and `lib/feedback-source/service.ts:227-238` both had every camelCase probe miss; their fixtures still pass unquoted names, a shape production cannot emit. One quoted-shape case each (`['"responseId"', '"tagId"']`, `['"surveyId"', '"elementId"']`) would pin the two user-visible bugs this PR fixed without claiming to.

- **`MCP_AUDIT_API_URL` drops the base path, and can silently record `localhost`.** `new URL("/api/mcp", WEBAPP_URL)` yields `https://host/api/mcp` for `WEBAPP_URL=https://host/formbricks`, because a leading `/` discards the base pathname. `getMcpResourceUrl()` (`modules/auth/lib/oauth-urls.ts:36`) already handles this via `appendPath`, so on a subpath deployment the audit record and the OAuth metadata would disagree about where the endpoint is. Separately, if `WEBAPP_URL` is unset in production the localhost fallback means every audit record permanently claims `http://localhost:3000/api/mcp` — valid URL, schema passes, silently wrong. Every other producer uses `req.url` and is immune; `authInfo.extra` already carries `requestId` and would be the natural channel for the real origin.

- **Module-scope `new URL` can crash at import time.** `WEBAPP_URL` is `z.url().optional()`, and zod accepts non-hierarchical schemes (`mailto:`, `urn:`, `foo:bar`) that `new URL(path, base)` rejects. That is a `TypeError` during evaluation of `mcp/constants.ts` → `mcp/server.ts` → `app/api/mcp/route.ts`: `next build` fails with a message that never names `WEBAPP_URL`, and at runtime the whole route's module graph is unloadable rather than one request failing. `oauth-urls.ts:13` does the same `new URL` lazily inside a function. Low likelihood, cheap to avoid.

- **The 403 `insufficient_scope` challenge still advertises the narrow scope** (`auth.ts:374`) while the 401 advertises the wide one. The RFC 6750 reading in the comment is defensible and I agree with it in the abstract — but ENG-2175's causal chain ("client derives its DCR scope from the challenge, authorizes against the wider metadata, gets `invalid_scope`") does not care which branch emitted the header, and this PR's own new "offline_access-only" test constructs exactly the token that lands here. Either use `MCP_CHALLENGE_SCOPE` on both branches, or keep the split and note in the comment that clients must re-register from the metadata document. Flagging rather than asserting, since the PR reasoned about this explicitly and reached the other conclusion.

- **`mapUniqueConstraintError` guesses on fallthrough.** `lib/feedback-source/service.ts:238` returns `FEEDBACK_SOURCE_NAME_DUPLICATE` for any unrecognised constraint, with no logging — so a user gets a confidently wrong message, renames the source, and hits the same error. Add an explicit `has("name")` guard and log the (PII-free) column names on fallthrough.

- **Comment/doc precision.** The `quote_identifier()` rule at `prisma-constraint.ts:21-23` is stated loosely: all-lowercase is necessary but not sufficient (`9lives`, `my-col` are quoted), the keyword category is wider than "reserved", and `quote_all_identifiers = on` quotes everything — which invalidates the "Postgres never quotes them" claims in two test comments. The "legacy `meta.target` is never quoted" claim also contradicts the PR's own test specifying behaviour for quoted legacy input. And `mcp-server.mdx:258` names `MCP_CHALLENGE_SCOPE` as the shared source when the root is `MCP_PROTECTED_RESOURCE_SCOPES`.

- **Three low-signal tests** could go: the idempotency case (entailed by the `token_hash` case), "logs the event when apiUrl is omitted" (byte-identical to an existing assertion at `service.test.ts:34`), and the matched-outer-pair case — which asserts on inputs Postgres cannot emit (`'"'`, `'""'`, `'"a"b"'`) while the input that would actually matter, the doubled form `'"a""b"'`, is absent. All academic, since Prisma column names cannot contain quotes.

- `auditLog ?? undefined` (`feedback-records.ts:95`) is dead narrowing — `buildV3AuditLog` already returns `TV3AuditLog | undefined`. And `mcp/constants.ts` now transitively imports `server-only` via `@/lib/constants`, so a future settings UI wanting `MCP_SERVER_NAME` would hit a confusing build error.

---

## What this PR does well

- `service.test.ts:94` reaching across to validate the real producer constant against the real consumer schema is the correct way to pin a cross-boundary contract, and is the one assertion here that would genuinely fail on regression.
- Replacing the literal scope mocks with `importOriginal` spreads in three test files — each annotated with the fact that *mocking the constants is what let ENG-2175 through* — fixes the bug class rather than the bug. Documenting a test's own blind spot (`surveys.test.ts:507`) is unusually honest and worth more than the test below it.
- The two integration tests drive real P2002s through real `adapter-pg` and assert the raw quoted meta *before* checking normalised output, so they cannot pass vacuously. That is the only way this bug class is catchable. (`response-error.integration.test.ts`'s second case omits its `rawFields` premise assertion, unlike the first — two lines to make it symmetric.)
- The `runMcpMutation` extraction is behaviour-identical line-for-line, including the error-path queue-then-rethrow. Inline ticket references match the established repo convention, each ID maps to exactly one fix, and no stale `MCP_MINIMUM_SCOPES` reference remains anywhere in the tree.

## Recommended order

1. Fix the `stamped` filter — the only change that makes production behaviour worse, and a two-line fix plus a real-`buildV3AuditLog` test.
2. Correct the four doc/comment inaccuracies (2-4).
3. Add the negative-control assertions (6) and de-tautologise the audit assertions (5).
4. Everything under Suggestions is follow-up material. The `MCP_API_ROUTE` / `MCP_AUDIT_API_URL` footgun and `logAuditEvent`'s unconditional swallow are worth a ticket of their own — this PR fixed the only offending call site, but nothing prevents the next one.

**Comment by xernobyl** [comment]
Thanks — the blocker was real, and it was this PR that would have made it live. All six numbered findings are fixed in the four commits above.

## 1 · The fabricated success audit events

Confirmed exactly as described. `buildAuditLogBaseObject` seeds `targetId: UNKNOWN_DATA` (`"unknown"`, truthy), only `createV3FeedbackRecords` overwrites it, and only for records it actually created — so the filter matched everything and the `// Nothing created` branch was unreachable.

Fixed by comparing against the placeholder. The point about *why* the tests could not catch it is the useful half: their mock omitted `targetId` entirely, so a truthy check behaved correctly in the test and wrongly in production. The mock now carries the placeholder, and I verified all three fail against the old filter — `expected 1 times, but got 2` on a batch where one record was created.

Agreed on the ordering, too: these events died in `logAuditEvent`'s `z.url()` check until now, so the `apiUrl` fix is what would have surfaced them. A false audit record is worse than a missing one for exactly the use case ENG-2173 exists for.

## 2 · Doc statements

All corrected, and there was one more than listed: the per-path spec repeats the draft-testability claim, so the handbook and the API reference were both wrong about it. Fixed in both. `openapi.yml` is regenerated via `pnpm api:v3:bundle`; the bundle diff is exactly the two description lines.

## 3 · "A superset is not enough"

You are right and I had it backwards. The provider filters `requestedScopes` against the registered set (`index.mjs:717`), so it is a subset test and the invariant is challenge ⊇ metadata. Reworded in the code comment and the test. We still assert equality, because both derive from one array — the stronger property, now stated as such rather than as the requirement.

## 4 · "mock-executes with side effects suppressed"

Removed from all three places. The accurate wording was already in the handler's own doc comment two lines up, which is a bit embarrassing. Your reasoning for why it matters — `isDryRun` being a real persisted concept, and "executing" being the wrong word next to a `read` gate — is why I rewrote rather than trimmed it.

## 5 · Self-referential audit assertions

Fixed. They now assert the shape (an absolute `…/api/mcp` URL) rather than the constant, and I verified they catch a regression to the bare path — five failures. The observation that this PR removes the same antipattern from `auth.test.ts` while adding it here is fair.

I did not take the branded type; `getMcpResourceUrl()` replacing the constant (below) removes the two-similar-names footgun at the source instead.

## 6 · Missing negative controls

Added: a `test.each` over all eight workflow write handlers, plus the missing `createV3SurveyResponse` assertion. Verified by lowering `patch` to `read` — exactly one test fails.

---

## From Suggestions, one promoted to a fix

**`MCP_AUDIT_API_URL` dropping the base path.** Confirmed: `new URL("/api/mcp", "https://host/formbricks")` yields `https://host/api/mcp`. Rather than reimplement `appendPath`, the call sites now use **`getMcpResourceUrl()`** — the same value the protected-resource metadata advertises, so the two agree by construction instead of coincidentally. That also makes it lazy, so the module-scope `new URL` can no longer throw during evaluation of the `/api/mcp` route graph, and `mcp/constants.ts` is back to zero imports and no longer transitively `server-only`. Three of your suggestions closed by one change.

Also done: `auditLog ?? undefined` dead narrowing removed, the `quote_identifier` rule stated accurately (all-lowercase is necessary but not sufficient; `quote_all_identifiers = on` quotes everything), and the integration test made symmetric — its second case now asserts the raw quoted premise like the first.

## Deliberately not done

- **`mapUniqueConstraintError` guessing on fallthrough** — real, but pre-existing and unrelated to the quoting fix. It belongs with the helper consolidation on ENG-1809, where I have left a comment covering it and the four parallel `isUniqueConstraintError` definitions.
- **The three low-signal tests** — I would rather keep the edge cases than trade them for a doubled-quote case that Prisma column names cannot produce.
- **The 403 `insufficient_scope` challenge staying narrow** — keeping the split. RFC 6750 `scope` names the scopes *required* for the resource, and `offline_access` is not one; a client that lands there has to re-register from the metadata regardless of which header it read. Called out in the comment so the next reader sees the choice rather than an oversight.

## Verification

7588 unit · 242 workflows · 77 integration · 0 TS errors · eslint clean.

Smoke-tested against a dev server on this branch: challenge ↔ metadata identical with a real DCR round-trip, duplicate `singleUseId` 200 → 409 on **both** v1 and v2, read-only key 200 on `/surveys/validate` and still 403 on `POST /surveys`.

Both of your authorization traces match what I found independently — worth recording that `validateV3Survey`'s only prisma-importing dependency keeps its write inside `ensureV3WorkspaceLanguages`, which the validate path never calls.

**Review by itsjavi — COMMENTED** [review]
A reminder to incorporate the change requested here about auth.test.ts https://github.com/formbricks/formbricks/pull/8718#pullrequestreview-4862637601

**Comment by xernobyl** [comment]
@itsjavi on it

**Review by Dhruwang — APPROVED** [review]
